File: app/src/main/java/ru/usedesk/sample/ui/screens/configuration/ConfigurationScreen.java
package ru.usedesk.sample.ui.screens.configuration;

import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import com.google.android.material.textfield.TextInputLayout;

import ru.usedesk.chat_sdk.UsedeskChatSdk;
import ru.usedesk.chat_sdk.entity.UsedeskChatConfiguration;
import ru.usedesk.common_sdk.entity.UsedeskEvent;
import ru.usedesk.knowledgebase_sdk.entity.UsedeskKnowledgeBaseConfiguration;
import ru.usedesk.sample.R;
import ru.usedesk.sample.databinding.ScreenConfigurationBinding;
import ru.usedesk.sample.model.configuration.entity.Configuration;
import ru.usedesk.sample.model.configuration.entity.ConfigurationValidation;

public class ConfigurationScreen extends Fragment {

    private ScreenConfigurationBinding binding;
    private ConfigurationViewModel viewModel;

    @NonNull
    public static ConfigurationScreen newInstance() {
        return new ConfigurationScreen();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.screen_configuration,
                container, false);

        viewModel = ViewModelProviders.of(this).get(ConfigurationViewModel.class);

        viewModel.getConfiguration()
                .observe(getViewLifecycleOwner(), this::onNewConfiguration);

        viewModel.getConfigurationValidation()
                .observe(getViewLifecycleOwner(), this::onNewConfigurationValidation);

        viewModel.getGoToSdkEvent()
                .observe(getViewLifecycleOwner(), this::onGoToSdkEvent);

        binding.btnGoToSdk.setOnClickListener(v -> onGoToSdk());

        binding.switchForeground.setOnCheckedChangeListener((buttonView, isChecked) -> {
            UsedeskChatSdk.stopService(requireContext());
        });

        try {
            String version = requireContext().getPackageManager()
                    .getPackageInfo(requireContext().getPackageName(), 0).versionName;
            binding.tvVersion.setText("v" + version);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        initTil(binding.tilUrlChat);
        initTil(binding.tilUrlOfflineForm);
        initTil(binding.tilUrlToSendFile);
        initTil(binding.tilUrlApi);
        initTil(binding.tilCompanyId);
        initTil(binding.tilAccountId);
        initTil(binding.tilToken);
        initTil(binding.tilClientEmail);
        initTil(binding.tilClientPhoneNumber);

        return binding.getRoot();
    }

    @Override
    public void onPause() {
        super.onPause();

        viewModel.setTempConfiguration(getConfiguration());
    }

    private void onGoToSdkEvent(@NonNull UsedeskEvent<Object> event) {
        event.process(data -> {
            ((IOnGoToSdkListener) getActivity()).goToSdk();
            return null;
        });
    }

    private void onGoToSdk() {
        viewModel.onGoSdkClick(getConfiguration());
    }

    private Configuration getConfiguration() {
        return new Configuration(
                binding.etUrlChat.getText().toString(),
                binding.etUrlOfflineForm.getText().toString(),
                binding.etUrlToSendFile.getText().toString(),
                binding.etUrlApi.getText().toString(),
                binding.etCompanyId.getText().toString(),
                binding.etAccountId.getText().toString(),
                binding.etToken.getText().toString(),
                binding.etClientSignature.getText().toString(),
                binding.etClientEmail.getText().toString(),
                binding.etClientName.getText().toString(),
                binding.etClientNote.getText().toString(),
                getLong(binding.etClientPhoneNumber.getText().toString()),
                getLong(binding.etClientAdditionalId.getText().toString()),
                binding.etClientInitMessage.getText().toString(),
                binding.etCustomAgentName.getText().toString(),
                binding.switchForeground.isChecked(),
                binding.switchKnowledgeBase.isChecked(),
                binding.switchWithSupportButton.isChecked());
    }

    @Nullable
    private Long getLong(@Nullable String value) {
        try {
            return Long.valueOf(value);
        } catch (Exception e) {
            return null;
        }
    }

    private void onNewConfiguration(@NonNull Configuration configuration) {
        binding.etUrlChat.setText(configuration.getUrlChat());
        binding.etUrlOfflineForm.setText(configuration.getUrlOfflineForm());
        binding.etUrlToSendFile.setText(configuration.getUrlToSendFile());
        binding.etUrlApi.setText(configuration.getUrlApi());
        binding.etCompanyId.setText(configuration.getCompanyId());
        binding.etAccountId.setText(configuration.getAccountId());
        binding.etToken.setText(configuration.getToken());
        binding.etClientSignature.setText(configuration.getClientSignature());
        binding.etClientEmail.setText(configuration.getClientEmail());
        binding.etClientName.setText(configuration.getClientName());
        String clientPhoneNumber;
        if (configuration.getClientPhoneNumber() == null) {
            clientPhoneNumber = "";
        } else {
            clientPhoneNumber = configuration.getClientPhoneNumber().toString();
        }
        binding.etClientPhoneNumber.setText(clientPhoneNumber);
        String clientAdditionalId;
        if (configuration.getClientAdditionalId() == null) {
            clientAdditionalId = "";
        } else {
            clientAdditionalId = configuration.getClientAdditionalId().toString();
        }
        binding.etClientAdditionalId.setText(clientAdditionalId);
        binding.etClientInitMessage.setText(configuration.getClientInitMessage());
        binding.etCustomAgentName.setText(configuration.getCustomAgentName());
        binding.switchForeground.setChecked(configuration.getForegroundService());
        binding.switchKnowledgeBase.setChecked(configuration.getWithKnowledgeBase());

        viewModel.getConfiguration().removeObservers(getViewLifecycleOwner());
    }

    private void initTil(TextInputLayout inputLayout) {
        inputLayout.getEditText().setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) {
                inputLayout.setError(null);
            }
        });
        inputLayout.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                inputLayout.setError(null);
            }
        });
    }

    private void showError(@NonNull TextInputLayout textInputLayout,
                           boolean isValid,
                           int errorStringId) {
        if (isValid) {
            textInputLayout.setError(null);
        } else {
            textInputLayout.setError(getResources().getString(errorStringId));
        }
    }

    private void onNewConfigurationValidation(
            @NonNull ConfigurationValidation configurationValidation
    ) {
        UsedeskChatConfiguration.Validation chatValidation =
                configurationValidation.getChatConfigurationValidation();

        showError(binding.tilUrlChat,
                chatValidation.getValidUrlChat(),
                R.string.validation_url_error);

        showError(binding.tilUrlOfflineForm,
                chatValidation.getValidUrlOfflineForm(),
                R.string.validation_url_error);

        showError(binding.tilUrlToSendFile,
                chatValidation.getValidUrlToSendFile(),
                R.string.validation_url_error);

        showError(binding.tilCompanyId,
                chatValidation.getValidCompanyId(),
                R.string.validation_empty_error);

        showError(binding.tilClientEmail,
                chatValidation.getValidClientEmail(),
                R.string.validation_email_error);

        showError(binding.tilClientPhoneNumber,
                chatValidation.getValidClientPhoneNumber(),
                R.string.validation_phone_error);

        UsedeskKnowledgeBaseConfiguration.Validation knowledgebaseValidation =
                configurationValidation.getKnowledgeBaseConfiguration();

        showError(binding.tilUrlApi,
                knowledgebaseValidation.getValidUrlApi(),
                R.string.validation_empty_error);

        showError(binding.tilAccountId,
                knowledgebaseValidation.getValidAccountId(),
                R.string.validation_empty_error);

        showError(binding.tilToken,
                knowledgebaseValidation.getValidToken(),
                R.string.validation_empty_error);
    }

    public interface IOnGoToSdkListener {
        void goToSdk();
    }
}

File: app/src/main/java/ru/usedesk/sample/ui/screens/configuration/ConfigurationScreen.kt

package ru.usedesk.sample.ui.screens.configuration

import android.content.pm.PackageManager
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.viewModels
import com.google.android.material.textfield.TextInputLayout
import ru.usedesk.chat_sdk.UsedeskChatSdk
import ru.usedesk.chat_sdk.UsedeskChatSdk.stopService
import ru.usedesk.common_gui.UsedeskFragment
import ru.usedesk.common_gui.UsedeskTextChangeListener
import ru.usedesk.common_gui.showInstead
import ru.usedesk.sample.GlideApp
import ru.usedesk.sample.R
import ru.usedesk.sample.databinding.ScreenConfigurationBinding
import ru.usedesk.sample.model.configuration.entity.Configuration
import ru.usedesk.sample.model.configuration.entity.ConfigurationValidation

class ConfigurationScreen : UsedeskFragment() {

    private val viewModel: ConfigurationViewModel by viewModels()
    private lateinit var binding: ScreenConfigurationBinding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(
            inflater,
            R.layout.screen_configuration,
            container,
            false
        )

        viewModel.modelFlow.onEachWithOld { old, new ->
            if (old?.avatar != new.avatar) {
                GlideApp.with(binding.ivAvatar)
                    .load(new.avatar)
                    .into(binding.ivAvatar)
            }
            if (old?.configuration != new.configuration) {
                new.configuration.onNewConfiguration()
            }
            if (old?.validation != new.validation) {
                new.validation?.onNewConfigurationValidation()
            }
            if (old?.clientToken != new.clientToken) {
                showInstead(binding.pbCreateChat, binding.btnCreateChat, new.clientToken.loading)
                new.clientToken.completed?.use {
                    binding.etClientToken.setText(it)
                    Toast.makeText(
                        requireContext(),
                        "Success:$it",
                        Toast.LENGTH_SHORT
                    ).show()
                }
                new.clientToken.error?.use {
                    Toast.makeText(
                        requireContext(),
                        "Failed:$it",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        }
        binding.etAvatar.addTextChangedListener(UsedeskTextChangeListener {
            viewModel.setAvatar(it)
        })
        binding.btnGoToSdk.setOnClickListener {
            val configuration = getConfiguration()
            if (viewModel.onGoSdkClick(configuration)) {
                (activity as IOnGoToSdkListener?)?.goToSdk(configuration)
            }
        }
        binding.btnCreateChat.setOnClickListener {
            val configuration = getConfiguration()
            if (viewModel.onCreateChat(configuration)) {
                val preparation = UsedeskChatSdk.initPreparation(
                    requireContext(),
                    configuration.toChatConfiguration()
                )
                viewModel.createChat(
                    preparation,
                    configuration.common.apiToken
                )
            }
        }
        binding.switchService.setOnCheckedChangeListener { _, _ ->
            stopService(requireContext())
        }
        binding.switchMaterialComponents.setOnCheckedChangeListener { _, _ ->
            if (viewModel.isMaterialComponentsSwitched(getConfiguration())) {
                requireActivity().finish()
                startActivity(requireActivity().intent)
            }
        }
        try {
            val version = requireContext().packageManager
                .getPackageInfo(requireContext().packageName, 0).versionName
            binding.tvVersion.text = "v$version"
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        binding.switchKbSection.setOnCheckedChangeListener { _, checked ->
            if (checked) {
                kbDeepLinkSwitched(section = checked)
            }
        }
        binding.switchKbCategory.setOnCheckedChangeListener { _, checked ->
            if (checked) {
                kbDeepLinkSwitched(category = checked)
            }
        }
        binding.switchKbArticle.setOnCheckedChangeListener { _, checked ->
            if (checked) {
                kbDeepLinkSwitched(article = checked)
            }
        }
        initTil(binding.tilUrlApi)
        initTil(binding.tilApiToken)
        initTil(binding.tilClientEmail)
        initTil(binding.tilUrlChat)
        initTil(binding.tilCompanyId)
        initTil(binding.tilChannelId)
        initTil(binding.tilClientPhoneNumber)
        initTil(binding.tilMessagesPageSize)
        initTil(binding.tilKbId)
        binding.ivAvatar.setOnClickListener {
            startImages()
        }
        registerFiles {
            it.firstOrNull()?.let { uri ->
                val url = uri.toString()
                binding.etAvatar.setText(url)
                viewModel.setAvatar(url)
            }
        }
        return binding.root
    }

    private fun kbDeepLinkSwitched(
        section: Boolean = false,
        category: Boolean = false,
        article: Boolean = false
    ) {
        binding.switchKbSection.isChecked = section
        binding.switchKbCategory.isChecked = category
        binding.switchKbArticle.isChecked = article
    }

    override fun onPause() {
        super.onPause()
        viewModel.setTempConfiguration(getConfiguration())
    }

    private fun getConfiguration(): Configuration {
        val additionalFields = mapOf(
            binding.etAdditionalField1Id.text.toString().trim() to
                    binding.etAdditionalField1Value.text.toString().trim(),
            binding.etAdditionalField2Id.text.toString().trim() to
                    binding.etAdditionalField2Value.text.toString().trim(),
            binding.etAdditionalField3Id.text.toString().trim() to
                    binding.etAdditionalField3Value.text.toString().trim()
        ).filter { it.key.isNotEmpty() }
            .map { it.key.toLong() to it.value }
            .toMap()
        val nestedFields = mapOf(
            binding.etAdditionalNestedField1Id.text.toString().trim() to
                    binding.etAdditionalNestedField1Value.text.toString().trim(),
            binding.etAdditionalNestedField2Id.text.toString().trim() to
                    binding.etAdditionalNestedField2Value.text.toString().trim(),
            binding.etAdditionalNestedField3Id.text.toString().trim() to
                    binding.etAdditionalNestedField3Value.text.toString().trim()
        ).filter { it.key.isNotEmpty() }
            .map { it.key.toLong() to it.value }
            .toMap()
        val additionalNestedFields = when {
            nestedFields.isEmpty() -> listOf()
            else -> listOf(nestedFields)
        }
        return Configuration(
            common = Configuration.Common(
                materialComponents = binding.switchMaterialComponents.isChecked,
                urlApi = binding.etUrlApi.text.toString(),
                apiToken = binding.etApiToken.text.toString(),
                clientEmail = binding.etClientEmail.text.toString(),
                clientName = binding.etClientName.text.toString(),
            ),
            chat = Configuration.Chat(
                urlChat = binding.etUrlChat.text.toString(),
                companyId = binding.etCompanyId.text.toString(),
                channelId = binding.etChannelId.text.toString(),
                messagesPageSize = binding.etMessagesPageSize.text.toString().toIntOrNull() ?: 1,
                clientToken = binding.etClientToken.text.toString(),
                clientNote = binding.etClientNote.text.toString(),
                clientPhoneNumber = binding.etClientPhoneNumber.text.toString().toLongOrNull(),
                clientAdditionalId = binding.etClientAdditionalId.text.toString(),
                clientInitMessage = binding.etClientInitMessage.text.toString(),
                clientAvatar = binding.etAvatar.text.toString(),
                customAgentName = binding.etCustomAgentName.text.toString(),
                messagesDateFormat = binding.etCustomDateFormat.text.toString(),
                messageTimeFormat = binding.etCustomTimeFormat.text.toString(),
                foregroundService = binding.switchService.isChecked,
                cacheFiles = binding.switchCacheFiles.isChecked,
                groupAgentMessages = binding.switchGroupAgentMessages.isChecked,
                adaptiveTimePadding = binding.adaptiveTimePadding.isChecked,
                additionalFields = additionalFields,
                additionalNestedFields = additionalNestedFields
            ),
            kb = Configuration.Kb(
                withKb = binding.switchKb.isChecked,
                withKbSupportButton = binding.switchKbWithSupportButton.isChecked,
                noBackStack = binding.switchKbNoBackStack.isChecked,
                kbId = binding.etKbId.text.toString(),
                sectionId = binding.etKbSectionId.text.toString().toLongOrNull(),
                section = binding.switchKbSection.isChecked,
                categoryId = binding.etKbCategoryId.text.toString().toLongOrNull(),
                category = binding.switchKbCategory.isChecked,
                articleId = binding.etKbArticleId.text.toString().toLongOrNull(),
                article = binding.switchKbArticle.isChecked
            )
        )
    }

    private fun Configuration.Common.onUpdated() {
        binding.switchMaterialComponents.isChecked = materialComponents
        binding.etUrlApi.setText(urlApi)
        binding.etApiToken.setText(apiToken)
        binding.etClientEmail.setText(clientEmail)
        binding.etClientName.setText(clientName)
    }

    private fun Configuration.Chat.onUpdated() {
        binding.etUrlChat.setText(urlChat)
        binding.etCompanyId.setText(companyId)
        binding.etChannelId.setText(channelId)
        binding.etMessagesPageSize.setText(messagesPageSize.toString())
        binding.etClientToken.setText(clientToken)
        binding.etClientPhoneNumber.setText(clientPhoneNumber?.toString() ?: "")
        binding.etClientAdditionalId.setText(clientAdditionalId ?: "")
        binding.etClientInitMessage.setText(clientInitMessage)
        binding.etCustomAgentName.setText(customAgentName)
        binding.etCustomDateFormat.setText(messagesDateFormat)
        binding.etCustomTimeFormat.setText(messageTimeFormat)
        binding.switchService.isChecked = foregroundService
        binding.switchCacheFiles.isChecked = cacheFiles
        binding.switchGroupAgentMessages.isChecked = groupAgentMessages
        binding.adaptiveTimePadding.isChecked = adaptiveTimePadding
        setAdditionalField(
            binding.etAdditionalField1Id,
            binding.etAdditionalField1Value,
            additionalFields,
            0
        )
        setAdditionalField(
            binding.etAdditionalField2Id,
            binding.etAdditionalField2Value,
            additionalFields,
            1
        )
        setAdditionalField(
            binding.etAdditionalField3Id,
            binding.etAdditionalField3Value,
            additionalFields,
            2
        )
        val nested = additionalNestedFields.firstOrNull() ?: mapOf()
        setAdditionalField(
            binding.etAdditionalNestedField1Id,
            binding.etAdditionalNestedField1Value,
            nested,
            0
        )
        setAdditionalField(
            binding.etAdditionalNestedField2Id,
            binding.etAdditionalNestedField2Value,
            nested,
            1
        )
        setAdditionalField(
            binding.etAdditionalNestedField3Id,
            binding.etAdditionalNestedField3Value,
            nested,
            2
        )
    }

    private fun Configuration.Kb.onUpdated() {
        binding.etKbId.setText(kbId)
        binding.etKbSectionId.setText(sectionId?.toString())
        binding.etKbCategoryId.setText(categoryId?.toString())
        binding.etKbArticleId.setText(articleId?.toString())
        binding.switchKb.isChecked = withKb
        binding.switchKbSection.isChecked = section
        binding.switchKbCategory.isChecked = category
        binding.switchKbArticle.isChecked = article
        binding.switchKbWithSupportButton.isChecked = withKbSupportButton
        binding.switchKbNoBackStack.isChecked = noBackStack
    }

    private fun Configuration.onNewConfiguration() {
        common.onUpdated()
        chat.onUpdated()
        kb.onUpdated()
    }

    private fun setAdditionalField(
        etId: EditText,
        etValue: EditText,
        fields: Map<Long, String>,
        index: Int
    ) {
        val id = fields.keys.toList().getOrNull(index)
        if (id != null) {
            val value = fields[id]
            etId.setText(id.toString())
            etValue.setText(value.toString())
        }
    }

    private fun initTil(inputLayout: TextInputLayout) {
        inputLayout.editText?.apply {
            onFocusChangeListener = OnFocusChangeListener { _: View?, hasFocus: Boolean ->
                if (hasFocus) {
                    inputLayout.error = null
                }
            }
            addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable) {
                    inputLayout.error = null
                }
            })
        }
    }

    private fun showError(
        textInputLayout: TextInputLayout,
        isValid: Boolean,
        errorStringId: Int
    ) {
        textInputLayout.error = when {
            isValid -> null
            else -> resources.getString(errorStringId)
        }
    }

    private fun ConfigurationValidation.onNewConfigurationValidation() {
        chatConfigurationValidation.run {
            showError(
                binding.tilUrlApi,
                validUrlApi,
                R.string.validation_url_error
            )
            showError(
                binding.tilUrlChat,
                validUrlChat,
                R.string.validation_url_error
            )
            showError(
                binding.tilCompanyId,
                validCompanyId,
                R.string.validation_empty_error
            )
            showError(
                binding.tilChannelId,
                validChannelId,
                R.string.validation_empty_error
            )
            showError(
                binding.tilClientEmail,
                validClientEmail,
                R.string.validation_email_error
            )
            showError(
                binding.tilClientPhoneNumber,
                validClientPhoneNumber,
                R.string.validation_phone_error
            )
        }

        knowledgeBaseConfiguration.run {
            showError(
                binding.tilUrlApi,
                validUrlApi,
                R.string.validation_url_error
            )
            showError(
                binding.tilApiToken,
                validToken,
                R.string.validation_empty_error
            )
            showError(
                binding.tilKbId,
                validKbId,
                R.string.validation_empty_error
            )
        }
    }

    interface IOnGoToSdkListener {
        fun goToSdk(configuration: Configuration)
    }
}